src/codegen/query_cache.cpp

```diff
@@ -0,0 +1,77 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// query_cache.cpp
+//
+// Identification: src/codegen/query_cache.cpp
+//
+// Copyright (c) 2015-17, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include "codegen/query_cache.h"
+#include "planner/delete_plan.h"
+#include "planner/insert_plan.h"
+#include "planner/seq_scan_plan.h"
+#include "planner/update_plan.h"
+#include "storage/data_table.h"
+
+namespace peloton {
+namespace codegen {
+
+Query* QueryCache::Find(const std::shared_ptr<planner::AbstractPlan> &key) {
+  auto it = cache_map_.find(key);
+  if (it == cache_map_.end()) {
+    return nullptr;
+  }
+  query_list_.splice(query_list_.begin(), query_list_, it->second);
+  return it->second->second.get();
+}
+
+void QueryCache::Add(const std::shared_ptr<planner::AbstractPlan> &key,
+                     std::unique_ptr<Query> &&val) {
+  query_list_.push_front(make_pair(key, std::move(val)));
+  cache_map_.insert(make_pair(key, query_list_.begin()));
```

Comment:
I think this function might be called by multiple worker threads at the same time. I may be wrong but It seems like this function is not thread-safe. Will it cause any problem?

Reply:
A good catch.  A read-write lock around `QueryCache` is added.